File: 0_1.baekjoon/graph/java/src/BOJ22868.java
import java.io.BufferedReader;
import java.io.InputStreamReader;
import java.util.ArrayList;
import java.util.Collections;
import java.util.HashMap;
import java.util.LinkedList;
import java.util.List;
import java.util.Map;
import java.util.Queue;
import java.util.StringTokenizer;
public class BOJ22868 {
    private static int N, M;
    private static Map<Integer, List<Integer>> graph;
    private static boolean[] path;
    private static int[] parent;
    public static void main(String[] args) throws Exception {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));

        StringTokenizer st = new StringTokenizer(br.readLine());

        graph = new HashMap<>();
        N = Integer.parseInt(st.nextToken());
        M = Integer.parseInt(st.nextToken());
        path = new boolean[N + 1];
        parent = new int[N + 1];
        for (int i=0;i<M;i++){
            st = new StringTokenizer(br.readLine());
            int u = Integer.parseInt(st.nextToken()), v = Integer.parseInt(st.nextToken());

            graph.putIfAbsent(u, new ArrayList<>());
            graph.putIfAbsent(v, new ArrayList<>());

            graph.get(u).add(v);
            graph.get(v).add(u);
        }

        st = new StringTokenizer(br.readLine());
        int s = Integer.parseInt(st.nextToken()), e = Integer.parseInt(st.nextToken());

        for (int key : graph.keySet()) {
            Collections.sort(graph.get(key));
        }

        for (int i=0;i<=N;i++) {
            parent[i] = i;
        }

        System.out.println(solve(s, e));

    }

    private static int solve(int s, int e) {
        int answer = bfs(s, e);

        pathCheck(e);
        path[s] = false;
        path[e] = false;

        return answer + bfs(e, s);
    }

    private static void pathCheck(int node) {
        path[node] = true;
        if (parent[node] == node) {
            return;
        }

        pathCheck(parent[node]);
    }

    private static int bfs(int s, int e) {
        Queue<int[]> q = new LinkedList<>();
        boolean[] visited = new boolean[N + 1];
        q.add(new int[] {s, 0});
        visited[s] = true;

        while (!q.isEmpty()) {
            int node = q.peek()[0];
            int length = q.poll()[1];

            for (int adj : graph.getOrDefault(node, new ArrayList<>())) {
                if (visited[adj] || path[adj]) continue;

                parent[adj] = node;

                if (adj == e) {
                    return length + 1;
                }

                visited[adj] = true;
                q.add(new int[] {adj, length + 1});
            }
        }

        return -1;
    }
}

File: 0_1.baekjoon/graph/kotlin/src/BOJ22868.kt
import java.util.*
import kotlin.collections.HashMap

private var N = 0
private var M = 0
private lateinit var graph: HashMap<Int, MutableList<Int>>
private lateinit var path: BooleanArray
private lateinit var parent: IntArray
fun main() = with(System.`in`.bufferedReader()) {
    val (n, m) = readLine().split(" ").map{ it.toInt()}
    N = n
    M = m

    graph = HashMap()
    path = BooleanArray(N + 1)
    parent = IntArray(N + 1)

    for (i in 0..N) {
        parent[i] = i
    }

    repeat(m) {
        val (u, v) = readLine().split(" ").map { it.toInt() }
        graph.putIfAbsent(u, mutableListOf())
        graph.putIfAbsent(v, mutableListOf())

        graph[u]!!.add(v)
        graph[v]!!.add(u)
    }

    for (key in graph.keys) {
        graph[key]!!.sort()
    }

    val (s, e) = readLine().split(" ").map { it.toInt() }

    print(solve(s, e))
}

private fun solve(s: Int, e: Int): Int {
    var result = bfs(s, e)
    checkPath(e)
    return result + bfs(e, s)
}

private fun checkPath(node: Int) {
    var node = node
    while(parent[node] != node) {
        path[node] = true
        node = parent[node]
    }
}

private fun bfs(s: Int, e: Int): Int {
    val q: Queue<Pair<Int, Int>> = LinkedList()
    val visited = BooleanArray(N + 1)
    q.add(Pair(s, 0))
    visited[s] = true

    while (q.isNotEmpty()) {
        val (node, length) = q.poll()

        for (adj in graph.getOrDefault(node, mutableListOf())) {
            if (visited[adj] || path[adj]) continue
            parent[adj] = node
            if (adj == e) {
                return length + 1
            }

            visited[adj] = true
            q.add(Pair(adj, length + 1))
        }
    }

    return -1
}